core/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceAggregatorBuilder.java

```diff
@@ -0,0 +1,236 @@
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ *
+ */
+public abstract class MultiValuesSourceAggregatorBuilder<VS extends ValuesSource, AB extends MultiValuesSourceAggregatorBuilder<VS, AB>>
+    extends ValuesSourceAggregatorBuilder<VS, AB> {
+
+    public static abstract class LeafOnly<VS extends ValuesSource, AB extends MultiValuesSourceAggregatorBuilder<VS, AB>>
+        extends MultiValuesSourceAggregatorBuilder<VS, AB> {
+
+        protected LeafOnly(String name, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) {
+            super(name, type, valuesSourceType, targetValueType);
+        }
+
+        /**
+         * Read from a stream that does not serialize its targetValueType. This should be used by most subclasses.
+         */
+        protected LeafOnly(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) throws IOException {
+            super(in, type, valuesSourceType, targetValueType);
+        }
+
+        /**
+         * Read an aggregation from a stream that serializes its targetValueType. This should only be used by subclasses that override
+         * {@link #serializeTargetValueType()} to return true.
+         */
+        protected LeafOnly(StreamInput in, Type type, ValuesSourceType valuesSourceType) throws IOException {
+            super(in, type, valuesSourceType);
+        }
+
+        @Override
+        public AB subAggregations(Builder subFactories) {
+            throw new AggregationInitializationException("Aggregator [" + name + "] of type [" +
+                type + "] cannot accept sub-aggregations");
+        }
+    }
+
+    private List<String> fields;
+    private Map<String, Script> scripts;
+
+    protected MultiValuesSourceAggregatorBuilder(String name, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) {
+        super(name, type, valuesSourceType, targetValueType);
+        if (fields == null) {
+            this.fields = Collections.emptyList();
+        }
+        if (scripts == null) {
+            this.scripts = Collections.emptyMap();
+        }
+    }
+
+    protected MultiValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType)
+        throws IOException {
+        super(in, type, valuesSourceType, targetValueType);
+    }
+
+    protected MultiValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType) throws IOException {
+        super(in, type, valuesSourceType);
+    }
+
+    /**
+     * Read from a stream.
+     */
+    @Override
+    protected void read(StreamInput in) throws IOException {
+        super.read(in);
+        if (in.readBoolean()) {
+            int size = in.readVInt();
+            fields = new ArrayList<>(size);
+            for (int i = 0; i < size; i++) {
+                fields.add(in.readString());
+            }
+        } else {
+            fields = Collections.emptyList();
```

Comment:
If you don't want to preserve null across the wire than just serialize the empty list on the way out and let readList read an empty list on the other side?